### PR TITLE
Issues

### LINKED ISSUE #569: "issues" and "quality checks"
When Changesets are created (either by FeedEater or by a person posting a new Changeset), they will be able to produce "issues" for further inspection.

Issue model:
- TL entity
- TL entity attribute
- FeedVersion
- description
- created by Changeset
- resolved by Changeset
- blocks Changeset application (boolean)

Open questions:
- should Issue-to-Entity be a 1-to-1 relationship? or have an EntityInIssue join model?
- how many of [Python validation library "quality checks"](https://github.com/google/transitfeed/tree/master/transitfeed) to include?
- if FeedEater level 1 changeset raises blocking issues, do we:

   a. automatically proceed to import SSPs, but with non-bound stop/route/operator Onestop IDs
   b. stop and let user trigger FeedEaterScheduleWorker after issue resolution
   c. always stop, even if no issues found, and have FeedEaterScheduleWorker as a second step

TODO:
- [ ] draw up sample data payloads, and decide how to relate issues and entities
- [ ] draw up list of "quality checks" that will produce issues
- ... actual implementation tasks

### PR DESCRIPTION
**Pull request by doublestranded**
closes #569

### CODE REVIEW

File: app/models/issue.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  has_many :entities_with_issues
+  belongs_to :created_by_changeset, class_name: 'Changeset'
+  belongs_to :resolved_by_changeset, class_name: 'Changeset'
+
+  validate :issue_types
+
+  ISSUE_TYPES = ['stop_position_inaccurate',
+                  'stop_rsp_distance_gap',
+                  'distance_calculation_inaccurate',
+                  'rsp_line_inaccurate',
```
Comment (doublestranded) [inline app/models/issue.rb]: Enumerize gem

File: app/models/changeset.rb [hunk 2]
```diff
@@ -154,8 +154,32 @@ def destroy_all_change_payloads
...
+    distance_rsps.merge(self.route_stop_patterns_created_or_updated)
+    distance_rsps.each { |rsp| rsp.calculate_distances }
+  end
+
+  def check_quality
+    gqc = QualityCheck::GeometryQualityCheck.new(changeset: self)
+    issues = []
+    issues += gqc.check
+    if self.imported_from_feed
+      import_score = ((gqc.distance_issue_tests - gqc.distance_issues).round(1)/gqc.distance_issue_tests).round(5) rescue 1.0
```
Comment (doublestranded) [inline app/models/changeset.rb]: move back into GeometryQualityCheck?

File: app/models/changeset.rb [hunk 3]
```diff
@@ -176,6 +200,12 @@ def apply!
           end
           EntityImportedFromFeed.import eiff_batch
         end
+
+        #computed_attributes unless self.imported_from_feed && self.imported_from_feed_version
+
+        issues = check_quality
+        return false, issues if issues.map(&:block_changeset_apply).include?(true)
+        issues.each(&:save!)
```
Comment (doublestranded) [inline app/models/changeset.rb]: Defer block changeset apply to later. Remove return false, issues line.

File: app/models/issue.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+#
+# Table name: issues
+#
+#  id                       :integer          not null, primary key
+#  created_by_changeset_id  :integer          not null
+#  resolved_by_changeset_id :integer
+#  details                  :string
+#  issue_type               :string
+#  open                     :boolean          default(TRUE)
+#  block_changeset_apply    :boolean          default(FALSE)
```
Comment (doublestranded) [inline app/models/issue.rb]: Remove block changeset apply for now.

File: app/services/quality_check.rb [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    self.distance_issue_tests = 0
+
+    distance_rsps = Set.new
+    stop_rsp_gap_pairs =  Set.new
+
+    # TODO if changeset is from import, we should go ahead and assign all rsps to distance_rsps
+
+    self.changeset.route_stop_patterns_created_or_updated.each do |rsp|
+      distance_rsps << rsp
+      rsp.stop_pattern.map { |onestop_id| OnestopId.find!(onestop_id) }.each do |stop|
```
Comment (doublestranded) [inline app/services/quality_check.rb]: build up list of onestop id's first (one query instead of multiple)

File: app/services/quality_check.rb [hunk 6]
```diff
@@ -0,0 +1,105 @@
+class QualityCheck
```
Comment (doublestranded) [inline app/services/quality_check.rb]: Define as module (as in OnestopId) with classes?

Comment (drewda) [inline app/services/quality_check.rb]: possibly relevant: http://stackoverflow.com/a/18737457/40956

File: app/controllers/api/v1/issues_controller.rb [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+  def create
+    entities_with_issues_params = issue_params.delete(:entities_with_issues).try(:compact)
+    issue = Issue.new(issue_params)
+    entities_with_issues_params.each { |ewi| issue.set_entity_with_issues_params(ewi) }
+    issue.created_by_changeset_id = issue_params[:created_by_changeset_id] || issue.changeset_from_entities.id
+
+    existing_issue = Issue.find_by_equivalent(issue)
+    if existing_issue
+      render json: existing_issue, status: :conflict
```
Comment (doublestranded) [inline app/controllers/api/v1/issues_controller.rb]: maybe redirect to show instead?

File: app/controllers/api/v1/issues_controller.rb [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+  def show
+    render json: @issue
+  end
+
+  def update
+    entities_with_issues_params = issue_params.delete(:entities_with_issues).try(:compact)
+    @issue.update!(issue_params)
+    entities_with_issues_params.each do |ewi|
+      @issue.set_entity_with_issues_params(ewi)
```
Comment (doublestranded) [inline app/controllers/api/v1/issues_controller.rb]: move set_entity_with_issues_params method into controller.

File: app/controllers/api/v1/issues_controller.rb [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+  end
+
+  private
+
+  def set_issue
+    @issue = Issue.find(params[:id])
+  end
+
+  def issue_params
+    params.require(:issue).permit!
```
Comment (drewda) [inline app/controllers/api/v1/issues_controller.rb]: http://edgeguides.rubyonrails.org/action_controller_overview.html#strong-parameters

Comment (doublestranded) [inline app/controllers/api/v1/issues_controller.rb]: list of acceptable param keys

File: app/controllers/api/v1/issues_controller.rb [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    end
+  end
+
+  def show
+    render json: @issue
+  end
+
+  def update
+    entities_with_issues_params = issue_params.delete(:entities_with_issues).try(:compact)
+    @issue.update!(issue_params)
```
Comment (doublestranded) [inline app/controllers/api/v1/issues_controller.rb]: What if wanted to remove individual entities with issues?

File: app/controllers/api/v1/issues_controller.rb (hunk 9 above)
Comment (drewda) [inline app/controllers/api/v1/issues_controller.rb]: have a look at strong params: http://edgeguides.rubyonrails.org/action_controller_overview.html#strong-parameters

File: app/models/json_schemas/changeset.json [hunk 11]
```diff
@@ -62,6 +62,9 @@
         "properties": {
           "action": {
             "type": { "enum": ["createUpdate", "destroy"] }
+          },
+          "issueResolved": {
```
Comment (doublestranded) [inline app/models/json_schemas/changeset.json]: Make this plural and array of numbers

File: app/models/change_payload.rb [hunk 12]
```diff
@@ -63,6 +63,9 @@ def apply!
     (payload_as_ruby_hash[:changes] || []).each do |change|
       (entity_types.keys & change.keys).each do |entity_type|
         changes << [entity_type, change[:action], change[entity_type]]
+        if change[:issue_resolved]
+          Issue.find(change[:issue_resolved]).update({ open: false, resolved_by_changeset: changeset})
```
Comment (doublestranded) [inline app/models/change_payload.rb]: !

**Comment by drewda** [comment]
@doublestranded two questions for you:

- can you add authentication to the issues controller actions that make changes? Can follow the same pattern as https://github.com/transitland/transitland-datastore/blob/master/app/controllers/api/v1/changesets_controller.rb#L6

- do you still get a namespace warning on `class QualityCheck`?

Otherwise, this looks ready to merge and test on dev, after @irees deploys station hierarchy code to prod.

**Pull request merged**